frame/support/src/migrations.rs

```diff
@@ -15,19 +15,126 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-#[cfg(feature = "try-runtime")]
-use crate::storage::unhashed::contains_prefixed_key;
 use crate::{
-	traits::{GetStorageVersion, NoStorageVersionSet, PalletInfoAccess, StorageVersion},
+	traits::{
+		GetStorageVersion, NoStorageVersionSet, OnRuntimeUpgrade, PalletInfoAccess, StorageVersion,
+	},
 	weights::{RuntimeDbWeight, Weight},
 };
 use impl_trait_for_tuples::impl_for_tuples;
 use sp_core::Get;
 use sp_io::{hashing::twox_128, storage::clear_prefix, KillStorageResult};
 use sp_std::marker::PhantomData;
+
 #[cfg(feature = "try-runtime")]
 use sp_std::vec::Vec;
 
+/// Make it easier to write versioned runtime upgrades.
+///
+/// VersionedRuntimeUpgrade allows developers to write migrations without worrying about checking
+/// and setting storage versions. Instead, the developer wraps their migration in this struct which
+/// takes care of version handling using best practices.
+///
+/// It takes 4 type parameters:
+/// - `Version`: The version being upgraded *from*.
```

Comment:
This is a good start, but I think we need `from` and `to`.  
It should advance exactly from `from` to `to`, so we can lock it down. No more `<` or `>` squishy logic anymore.

Reply:
I originally implemented this with `From` and `To`, but refactored to only `Version` after thinking that `To` is redundant.

I see `To` as only useful to support these two cases, which I can't think of scenarios where they would be best practice:
1. Skipping version numbers (even the nom pools implemented `MigrateV3ToV5` using two migrations, each incrementing one version: https://github.com/paritytech/substrate/blob/aeb8c3107f79afef081dde2b9aa2866a6596f6ba/frame/nomination-pools/src/migration.rs#L497 )
2. A single migration running on more than one version - this seems strange, and probably means the migration should be rearchitected

Let me know if you can think of any scenarios I've missed, or if you think we should support any of the above cases, and I'm happy to switch back to `To`/`From`.

Reply:
> 2\. A single migration running for subsequent upgrades - this seems strange and unnecessary. I can't think of cases where it would be needed.

what if somebody wants to implement a migration from version 2 to version 4 in a more efficient way than executing version 2 to 3 and version 3 to 4 ?
Also we could provide different type. One type for upgrade from one version to next version, one type for upgrade from one version to another version.

Reply:
Yes we actually have this in the scheduler: `1->4`, `2->4` and nom pools also had a batched migration from `3->5`.  
I think making this explicit also helps with chaining them together later on.

>One type for upgrade from one version to next version, one type for upgrade from one version to another version.

Yea the former could be a wrapper around the latter.